File: 2021/5/main.py
from collections import defaultdict
s = [i.split(" -> ") for i in open("input.txt").read().split("\n")]
start = []
end = []
for i, j in s:
    i = [int(x) for x in i.split(",")]
    j = [int(x) for x in j.split(",")]
    start.append(i)
    end.append(j)
p2 = True
d = defaultdict(int)
for i in range(len(start)):
    if start[i][0] == end[i][0]:
        for j in range(min(start[i][1], end[i][1]), max(start[i][1], end[i][1]) + 1):
            d[(start[i][0]), j] += 1
    if start[i][1] == end[i][1]:
        for j in range(min(start[i][0], end[i][0]), max(start[i][0], end[i][0]) + 1):
            d[(j, start[i][1])] += 1
    if p2:
        if abs(start[i][0] - end[i][0]) == abs(start[i][1] - end[i][1]):
            if start[i][0] > end[i][0]:
                dx = -1
            else:
                dx = 1
            if start[i][1] > end[i][1]:
                dy = -1
            else:
                dy = 1
            x, y = start[i]
            while(x != end[i][0] and y != end[i][1]):
                d[(x, y)] += 1
                x += dx
                y += dy
            d[(x, y)] += 1
        

print(sum(i > 1 for i in d.values()))
    

File: 2021/5/main.jl
using DataStructures

function p1()
    d = DataStructures.DefaultDict{Pair{Int, Int}, Int}(0)
    for line in readlines("input.txt")
        s, e = [parse.(Int, split(i, ',')) for i in split(line, " -> ")]
        if s[1] == e[1]  # Vertical line
            for i in min(s[2], e[2]):max(s[2], e[2])
                d[Pair(s[1], i)] += 1
            end
        elseif s[2] == e[2]  # Horizontal line
            for i in min(s[1], e[1]):max(s[1], e[1])
                d[Pair(i, s[2])] += 1
            end
        end
    end
    println(sum((item -> item.second >= 2), d))
end

function p2()
    d = DataStructures.DefaultDict{Pair{Int, Int}, Int}(0)
    for line in readlines("input.txt")
        s, e = [parse.(Int, split(i, ',')) for i in split(line, " -> ")]
        if s[1] == e[1]  # Vertical line
            for i in min(s[2], e[2]):max(s[2], e[2])
                d[Pair(s[1], i)] += 1
            end
        elseif s[2] == e[2]  # Horizontal line
            for i in min(s[1], e[1]):max(s[1], e[1])
                d[Pair(i, s[2])] += 1
            end
        elseif abs(s[1] - e[1]) == abs(s[2] - e[2])
            dx = s[1] > e[1] ? -1 : 1
            dy = s[2] > e[2] ? -1 : 1
            cx, cy = s
            while cx != e[1]
                d[Pair(cx, cy)] += 1
                cx += dx
                cy += dy
            end
            d[Pair(cx, cy)] += 1
        end
    end
    println(sum((item -> item.second >= 2), d))
end
p1()
p2()
